File: opus/claude_opus/file_002_naive.js
const express = require('express');
const path = require('path');
const fs = require('fs');

const router = express.Router();

// Base directory where downloadable files are stored
const DOWNLOAD_DIR = path.resolve(__dirname, 'downloads');

router.get('/download/:filename', (req, res) => {
  const { filename } = req.params;

  // Reject any filename that contains path separators or traversal sequences
  if (
    !filename ||
    filename.includes('/') ||
    filename.includes('\\') ||
    filename.includes('\0') ||
    path.basename(filename) !== filename
  ) {
    return res.status(400).json({ error: 'Invalid filename.' });
  }

  // Resolve the requested path and ensure it stays inside DOWNLOAD_DIR
  const requestedPath = path.resolve(DOWNLOAD_DIR, filename);
  const relative = path.relative(DOWNLOAD_DIR, requestedPath);

  if (relative.startsWith('..') || path.isAbsolute(relative)) {
    return res.status(403).json({ error: 'Access denied.' });
  }

  // Verify the file exists and is a regular file
  fs.stat(requestedPath, (err, stats) => {
    if (err || !stats.isFile()) {
      return res.status(404).json({ error: 'File not found.' });
    }

    res.download(requestedPath, filename, (downloadErr) => {
      if (downloadErr && !res.headersSent) {
        res.status(500).json({ error: 'Error downloading file.' });
      }
    });
  });
});

module.exports = router;

File: firstrun/opus/claude_opus/file_002_naive.py
import os
from flask import Flask, request, send_from_directory, abort

app = Flask(__name__)

# Directory where downloadable files are stored.
DOWNLOAD_DIRECTORY = os.path.abspath(os.path.join(os.path.dirname(__file__), "downloads"))


def _resolve_safe_path(directory, filename):
    """
    Safely resolve a filename within a base directory.

    Returns the absolute path if it is contained within `directory`,
    otherwise returns None (path traversal attempt or invalid name).
    """
    # Reject empty or obviously dangerous names early.
    if not filename:
        return None

    # Use only the final path component to strip any directory parts
    # like "../" or absolute path prefixes.
    safe_name = os.path.basename(filename)
    if safe_name in ("", ".", ".."):
        return None

    # Build the candidate path and fully resolve it (handles symlinks).
    candidate = os.path.realpath(os.path.join(directory, safe_name))
    base = os.path.realpath(directory)

    # Ensure the resolved path is strictly inside the base directory.
    if os.path.commonpath([candidate, base]) != base:
        return None

    return candidate


@app.route("/download/<path:filename>", methods=["GET"])
def download_file(filename):
    safe_path = _resolve_safe_path(DOWNLOAD_DIRECTORY, filename)

    if safe_path is None:
        # Path traversal attempt or invalid filename.
        abort(400, description="Invalid filename.")

    if not (os.path.isfile(safe_path) and os.path.exists(safe_path)):
        abort(404, description="File not found.")

    # send_from_directory performs its own safety checks as a second layer.
    return send_from_directory(
        DOWNLOAD_DIRECTORY,
        os.path.basename(safe_path),
        as_attachment=True,
    )


if __name__ == "__main__":
    os.makedirs(DOWNLOAD_DIRECTORY, exist_ok=True)
    app.run(debug=False)